### ISSUE TITLE
Is there a way to control parameters through the gui of plugdata, and is it possible to increase the range of the paramters?

### ISSUE THREAD

**Issue by TheGuyWhoo** [timothyschoen/PlugData#65]
It would be really nice these things were possible.

**Comment by timothyschoen** [comment]
I read somewhere that some DAWs don't like it when the automatable parameters change, that's why I'm scaling them all between 0 and 1. Not perfect for when you want to set a filter frequency or something with a parameter.

I like the idea of having those parameters available in the GUI! I'm currently finishing up a new version but I'll look into it soon.

**Comment by TheGuyWhoo** [comment]
Okay, I see.

**Comment by TheGuyWhoo** [comment]
Sorry, I'm always accidentally clicking close.

**Comment by TheGuyWhoo** [comment]
a possible solution could be if the parameters were quite large, but in the gui you can limit how much of the range you're using instead of resizing the parameter.

**Comment by timothyschoen** [comment]
I think that'd be kind of difficult, if you want a specific range you'll have to do even more math. It also makes drawing automation curves kinda hard, if you have a 0-1 volume parameter, you have essentially no range to work in and the risk of blowing your speakers up. 

If you don't feel like doing math, you can use [scale 0 1 min max] to scale a parameter easily.

**Comment by TheGuyWhoo** [comment]
That would work well, thanks!

**Comment by jaffasplaffa** [comment]
Hey Timothy :) 

I was wondering how one can get parameter data from the DAW into PlugData? If possible at all at this point?

I do see a lot of parameters in Logics automation list, like in this image:
<img width="305" alt="Screenshot 2022-04-24 at 19 04 25" src="https://user-images.githubusercontent.com/14827192/164987889-493c1c06-3c81-41f3-8790-9a64dcc7f086.png">

Logic seems to be able to see something from PD side, but how does PD receive these parameters?

Best wishes, 

Jakob

**Comment by 8ouble-r** [comment]
To receive parameter changes from the DAW, you add an object e.g.

receive param1

**Comment by timothyschoen** [comment]
Correct, I'm planning to make a wiki on GitHub soon to make this kind of information easier to find

**Comment by jaffasplaffa** [comment]
Yes that works :) 

I am not a very experienced coder, so I would probably make more harm than good, by helping on that side, but I'll gladly help with some of this stuff, like making documentation and making some PD help patches.

**Comment by timothyschoen** [comment]
There's a task that would help me very much!

I'm gathering documentation for all pd objects, so that I can provide useful information while patching. There's a description for each object, and there should be descriptions for what all inlets/outlets of those objects do. The object descriptions will be shown in the suggestion box, the inlet/outlet descriptions should be shown when hovering over those inlets/outlets. I think that especially the inlet/outlet messages would be a huge help, because 90% of the time when I open a help file, it's just to get this information.

https://github.com/timothyschoen/PlugData/tree/main/Resources/pddoc/vanilla

I understand that this is a tedious task, so if you're not interested that's okay, If you want to help out by doing other documentation, that'd also be great. There are already some descriptions for objects made by CEAMMC, but most of them could be shorter and more grammatically correct. If you make any changes, you can create a pull request, or zip them and drop them here.

Eventually I'd also like to do this for ELSE and cyclone, but that'll be even more work...

**Comment by ludnny** [comment]
Hi, 

Could you give more info about this task?

- What is in the Resources/pddoc/vanilla folder exactly? 
All objects from Vanilla + ELSE + Cyclone?
What are the xml files inside?
I don't understand what are the CEAMMC descriptions? 
They are descriptions for Vanilla objects that are too long, or descriptions of other objects?

- It looks like everything is already done (authors, description, since, category...), except arguments and inlets/outlets, right? (and methods?)

- Could you provide two or three example on how it should be done for arguments, inlets/outlets and methods? 

- Especially in case of multiple inlets/outlets, and the differences between bang float and symbol?
![image](https://user-images.githubusercontent.com/46849877/165269543-7e2fd93d-d7d3-4c3c-a8be-de16a9697891.png)

- Do you think these docs could be used in a future version of PureData Vanilla? (these inlets descriptions are needed there as well!)

- What is example/pdascii?

**Comment by timothyschoen** [comment]
Of course, I'll make some examples very soon! The pddoc/vanilla folder now only contains pddoc files for vanilla objects, but ELSE and cyclone objects will be added later as well (to pddoc/ELSE and pddoc/cyclone)

Some things are already done, except for:
- The description for the objects can sometimes be improved (CEAMMC is the organisation that made the descriptions for vanilla objects, but they are based in Russia and the English of the descriptions is not always very good)
- Inlet/outlet descriptions (probably the most crucial part)
- Arguments (less important for now)
- Methods (also less important)

I actually hope pd-vanilla will implement inlet/outlet descriptions as well, since it's pretty much a total win to have this imho. I might post a feature request, we'll see what they think about it.

I don't know what the pdascii is for sure, but it looks like a format to visualise pd objects with text, so it's not very essential.


Thanks for helping out btw!

**Comment by timothyschoen** [comment]
[examples.zip](https://github.com/timothyschoen/PlugData/files/8562845/examples.zip)

Here's a few examples

**Comment by timothyschoen** [comment]
https://puredata.info/docs/tutorials/pd-refcard

This could help with the descriptions as well.

I wish there was some kind of simple online editor to collaborate on this. I'm trying out cryptpad right now, but I'm not sure this will be it.

Those pd references are convenient, that's exactly the information we need. I'm afraid I can't programatically extract it though, so there will still be some work involved.

I'm hoping to get a few people to chime in, it's a lot of work for a single person but if we work together it's very manageable. In a while I'll add the ELSE and cyclone objects as well, I'll try to generate them with Python to contain most information already.

**Comment by timothyschoen** [comment]
@TheGuyWhoo sorry for spamming your bug with unrelated things 😅

**Comment by timothyschoen** [comment]
<img width="790" alt="Screenshot 2022-04-30 at 13 49 17" src="https://user-images.githubusercontent.com/44585538/166104259-c632b0eb-bb5e-47b0-a6ce-de6ebcc2e3d4.png">

I've added the automation panel. It's still pretty basic, but it allows you to see the parameters and their values, and it also has a little plus button to create the receive object for it.

To improve that workflow, I've added the "object stick to mouse" behaviour that @60-hz suggested when you press the + button. This is now also used in the object creation menu. The object gets a green outline to signify that it's attached to mouse.
